### PR TITLE
Use external transport crate #366

### PR DESCRIPTION
**Pull request by gijsvl**
The extracted transport crate can be found here: https://github.com/boltlabs-inc/transport

We should make sure it all still makes sense, including the level of abstraction.
Specifically, the feature `allow_explicit_certificate_trust`, does that make sense to also live inside the transport crate, I do believe so, but just checking. Also, the abstraction for `ZkChannelAddress` that now implements the trait `Address`, which is part of the transport crate.

### CODE REVIEW

File: src/transport.rs
Comment (marsella) [inline src/transport.rs:11]: We have been using a more consolidated use structure in most of the crate, but it has to be done manually. Maybe we can find a way to explicitly fix this via `fmt`. These lines would be

```
use crate::{customer, transport::...};
use {
  http::{uri::InvalidUri, Uri},
  std::{fmt::{self, Display}, str::FromStr},
  thiserror::Error,
  ...
}
```

I think the benefits are consolidation (easier to scan), automatic alphabetization by fmt.

File: src/customer.rs
Comment (marsella) [inline src/customer.rs:15]: Do these have to say `use X::self as X`? I thought if you keep the same name, you could just use `self` and it would automatically use the X as the name.

File: src/transport.rs
Comment (gijsvl) [reply src/transport.rs:11]: I almost never look at the imports (except when I'm having some conflict), so I'm okay with whatever style we choose. However, if it doesn't happen automatically (preferably cargo fmt, maybe just IDE), I'm very much opposed to any choice of style, because that would mean manual work. There is this unstable rustfmt feature: https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#imports_granularity, I believe what you are suggesting is somewhat close to the "Crate" option, although it might be even closer to the "One" option. Unfortunately, the "One" option doesn't seem to work at this point. We can discuss what option seems most reasonable, or if there is another way to enforce it.

File: src/lib.rs
Comment (gijsvl) [reply src/lib.rs:13]: At some point it seemed to matter, but that might have been some intermediary state. I'll remove it.

File: src/customer.rs
Comment (gijsvl) [reply src/customer.rs:15]: apparently not 😇

File: src/lib.rs
Comment (gijsvl) [reply src/lib.rs]: Ah, I remember, it is needed because the `ZkChannelAddress` is used inside the documentation, hence, it needs to be public. Note that the transport mentioned here is merely that object and not much else (everything contained in `transport.rs`.

**Comment by gijsvl** [comment]
With regard to the transport crate, indeed, not much changed, it's basically the same with the exception of the abstract Trait `Address`, as you've mentioned.

File: src/transport.rs
Comment (marsella) [reply src/transport.rs:11]: Good feature. Per discussion offline, I'd be happy with `Crate` level formatting - I'm not particularly tied to what we currently have. The [tracking issue](https://github.com/rust-lang/rustfmt/issues/4991) implies that there are stability problems but doesn't cite any specific ones other than losing comments in the includes, which we don't have, and [there's an anecdote](https://github.com/LemmyNet/lemmy/blob/main/.rustfmt.toml) that `Crate`-level formatting has been stable for a while.

Since this repo already uses nightly rust, I think you can add a `.rustfmt.toml` file like the one linked above and specify this. I would expect that formatting config will automatically propagate to CI.

File: src/lib.rs
Comment (marsella) [reply src/lib.rs]: I see, that seems reasonable. `ZkChannelAddress` is something that external users might have to deal with anyway (e.g. to format their RPC requests eventually).

**Review by marsella — COMMENTED** [review]
Thank you for the fixes. My only remaining recommendation is to update the `rustfmt.toml` with the `imports_granularity` specification - it looks like that file already supports unstable features, although I'm not sure which ones. Mukund may have mentioned something about line-wrapping documentation, but I think that is an issue for a different PR.

**Comment by gijsvl** [comment]
Before merging, we should merge https://github.com/boltlabs-inc/transport/pull/1, and change the transport dependency back to "main" branch

**Review by marsella — APPROVED** [review]
Looks good, approving pending the transport fix.

**Pull request merged**
